### PR TITLE
Fix a few static linking issues

### PR DESCRIPTION
**Pull request by am11**
Changes:

* Define `__xmknod` as `mknod`, because libSystem.Native.a in official package is compiled on machine with older version of glibc (where `mknod` was `__xmknod`) and newer glibc doesn't have that symbol. It's pretty much a nop if it's unused.
* Use the same compiler to build static libs which is selected for final linkage. Defer the cmake calls until after the toolchain selection.
* Add symbol prefix only on osx.
* To make gcc happy, `libm` must come after icu libs; moved it to the end of libs list.
* Move -static near -shared (so it's easier to track in logs where to expect it).

Notes:
* static executable works if we publish with `-p:PositionIndependentExecutable=false` (aka without `-static-pie`), otherwise it runs into sigsegv at the very beginning (during `__init_libc` on Alpine and it was crashing the debugger on Ubuntu when I stopped trying..)
* publishing static executable require the same toolchains, both compiler and linker are of gcc-toolchaiin or llvm-toolchain (with `-p:UseLlvmLinker=true`). Mixed toolchain builds are running  into different linker errors on different distros.

### CODE REVIEW

File: src/installer/pkg/projects/Microsoft.DotNet.ILCompiler/Microsoft.DotNet.ILCompiler.pkgproj [hunk 1]
```diff
@@ -17,11 +17,11 @@
...
-      <File Include="$(MSBuildThisFileDirectory)\..\..\..\..\native\libs\System.Security.Cryptography.Native\*" TargetPath="native/src/libs/System.Security.Cryptography.Native"/>
-      <File Include="$(MSBuildThisFileDirectory)\..\..\..\..\native\libs\build-local.sh" TargetPath="native/src/libs/build-local.sh"/>
-      <File Include="$(MSBuildThisFileDirectory)\..\..\..\..\native\minipal\*" TargetPath="native/src/minipal"/>
-      <File Include="$(MSBuildThisFileDirectory)\..\..\..\..\native\libs\Common\*" TargetPath="native/src/libs/Common"/>
+      <File Include="$(SharedNativeRoot)libs\System.Globalization.Native\*" TargetPath="native/src/libs/System.Globalization.Native"/>
+      <File Include="$(SharedNativeRoot)native\libs\System.Security.Cryptography.Native\*" TargetPath="native/src/libs/System.Security.Cryptography.Native"/>
+      <File Include="$(SharedNativeRoot)native\libs\build-local.sh" TargetPath="native/src/libs/build-local.sh"/>
+      <File Include="$(SharedNativeRoot)native\minipal\*" TargetPath="native/src/minipal"/>
+      <File Include="$(SharedNativeRoot)native\libs\Common\*" TargetPath="native/src/libs/Common"/>
```
Comment (am11) [inline src/installer/pkg/projects/Microsoft.DotNet.ILCompiler/Microsoft.DotNet.ILCompiler.pkgproj:24]:
```suggestion
      <File Include="$(SharedNativeRoot)libs\System.Security.Cryptography.Native\*" TargetPath="native/src/libs/System.Security.Cryptography.Native"/>
      <File Include="$(SharedNativeRoot)libs\build-local.sh" TargetPath="native/src/libs/build-local.sh"/>
      <File Include="$(SharedNativeRoot)minipal\*" TargetPath="native/src/minipal"/>
      <File Include="$(SharedNativeRoot)libs\Common\*" TargetPath="native/src/libs/Common"/>
```

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets [hunk 2]
```diff
@@ -142,8 +139,9 @@ The .NET Foundation licenses this file to you under the MIT license.
       <LinkerArg Include="-Wl,-z,relro" Condition="'$(TargetOS)' != 'osx'" />
       <!-- binskim warning BA3011 The BIND_NOW flag is missing -->
       <LinkerArg Include="-Wl,-z,now" Condition="'$(TargetOS)' != 'osx'" />
-      <LinkerArg Include="-Wl,-u,_NativeAOT_StaticInitialization" Condition="('$(UseLLVMLinker)' == 'true' or '$(TargetOS)' == 'osx' or '$(TargetOS)' == 'freebsd') and '$(NativeLib)' == 'Shared'" />
+      <LinkerArg Include="-Wl,-u,$(_SymbolPrefix)NativeAOT_StaticInitialization" Condition="('$(UseLLVMLinker)' == 'true' or '$(TargetOS)' == 'osx' or '$(TargetOS)' == 'freebsd') and '$(NativeLib)' == 'Shared'" />
       <LinkerArg Include="-Wl,--require-defined,NativeAOT_StaticInitialization" Condition="'$(UseLLVMLinker)' != 'true' and '$(TargetOS)' == 'linux' and '$(NativeLib)' == 'Shared'" />
+      <LinkerArg Include="-Wl,--defsym,__xmknod=mknod" Condition="'$(StaticExecutable)' == 'Static'" />
```
[the remark this answers is not in the document]
Comment (am11) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:144]: Sure, per https://github.com/bminor/glibc/commit/99468ed45f5a58f584bab60364af937eb6f8afda, it was 2.33. In Ubuntu 22.04, there is glibc 2.35.

[the remark this answers is not in the document]
Comment (am11) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:144]:
```suggestion
      <!-- this workaround can be deleted once the minimum supported glibc version
           (runtime's official build machine's glibc version) is at least 2.33
           see https://github.com/bminor/glibc/commit/99468ed45f5a58f584bab60364af937eb6f8afda -->
      <LinkerArg Include="-Wl,--defsym,__xmknod=mknod" Condition="'$(StaticExecutable)' == 'true'" />
```

**Comment by am11** [comment]
> static executable works if we publish with `-p:PositionIndependentExecutable=false` (aka without `-static-pie`), otherwise it runs into sigsegv at the very beginning (during `__init_libc` on Alpine and it was crashing the debugger on Ubuntu when I stopped trying..)

@jkotas, would it make sense to invert `PositionIndependentExecutable` logic for `-static-pie`?

```diff
-      <LinkerArg Include="-static-pie" Condition="'$(StaticExecutable)' == 'true' and '$(PositionIndependentExecutable)' != 'false'" />
+      <LinkerArg Include="-static-pie" Condition="'$(StaticExecutable)' == 'true' and '$(PositionIndependentExecutable)' == 'true'" />
```
The executable doesn't run without `-p:PositionIndependentExecutable=false` during publish (at least) on linux-arm64 and linux-musl-arm64 where I was testing today.

**Comment by jkotas** [comment]
We generally try to have secure defaults. PositionIndependentExecutable=true is the secure default.

Is `-static-pie` broken because of something we do, or is it universally broken? In other words, is a simple "Hello world" C app broken with `-static-pie` on Linux arm64 as well?

**Comment by am11** [comment]
> Is `-static-pie` broken because of something we do

I think it is something we do.

```sh
$ echo "int main(void) { return 42; }" | clang -xc -static -static-pie -

$ ./a.out

$ echo $?
42

$ file ./a.out
./a.out: ELF 64-bit LSB pie executable, ARM aarch64, version 1 (GNU/Linux), static-pie linked, BuildID[sha1]=559bd6a34b13388f80de1a33c6222e79d4aee185, for GNU/Linux 3.7.0, not stripped
```

On the same machine, .NET 8 with these changes:
```sh
$ apt install -y clang lld libz-dev build-essential nano cmake curl libicu-dev libssl-dev libnuma-dev

# clang with llvm linker
$ dotnet publish -c Release -o dist \
    -p:'PublishAot=true;StaticExecutable=true;StaticIcuLinking=true;StaticOpenSslLinking=true;StaticNumaLinking=true' \
    -p:UseLlvmLinker=true

$ dist/exe1          
Segmentation fault

# clang with gnu linker (no difference)
$ rm -rf obj bin dist
$ dotnet publish -c Release -o dist \
    -p:'PublishAot=true;StaticExecutable=true;StaticIcuLinking=true;StaticOpenSslLinking=true;StaticNumaLinking=true'

$ dist/exe1          
Segmentation fault
```
With `-p:PositionIndependentExecutable=false` and clang on Ubuntu, we get a pure static executable which works (I'm testing a simple Brotli compression sample program). However, even disabling -static-pie does not work with gcc and the published executable aborts at startup. With -static-pie, gcc throws link-time error with both gnu-ld and llvm-ld.

Alpine is a different story: with clang and llvm-ld OR gcc and gnu-ld, the app works without -static-pie (mixing toolchains run into issues). Without -static-pie all combinations abort, even that `./a.out` is aborting there with static-pie.

Still, `statically linked` vs. `static-pie linked` have a few differences even how ldd sees them and keeping static-pie for StaticExecutable is a good default, I agree. However, currently a simple hello world app is also running into problems. It needs more work, that's for sure. :)

**Review by jkotas — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**